cmd/occollector/app/builder/builder.go

```diff
@@ -0,0 +1,138 @@
+package builder
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/spf13/viper"
+)
+
+const (
+	receiversRoot   = "receivers"
+	jaegerEntry     = "jaeger"
+	opencensusEntry = "opencensus"
+	zipkinEntry     = "zipkin"
+)
+
+// JaegerReceiverCfg holds configuration for Jaeger receivers.
+type JaegerReceiverCfg struct {
+	// JaegerThriftTChannelPort is the port that the relay receives on for jaeger thrift tchannel requests
+	JaegerThriftTChannelPort int `mapstructure:"thrift-tchannel-port"`
+	// ReceiverJaegerHTTPPort is the port that the relay receives on for jaeger thrift http requests
+	JaegerThriftHTTPPort int `mapstructure:"thrift-http-port"`
+}
+
+// JaegerReceiverEnabled checks if the Jaeger receiver is enabled, via a command-line flag, environment
+// variable, or configuration file.
+func JaegerReceiverEnabled(v *viper.Viper, cmdFlag string) bool {
+	return featureEnabled(v, cmdFlag, receiversRoot, jaegerEntry)
+}
+
+// NewJaegerReceiverCfg returns an instance of JaegerReceiverCfg with default values
+func NewJaegerReceiverCfg() *JaegerReceiverCfg {
+	opts := &JaegerReceiverCfg{
+		JaegerThriftTChannelPort: 14267,
+		JaegerThriftHTTPPort:     14268,
+	}
+	return opts
+}
+
+// InitFromViper returns a JaegerReceiverCfg according to the configuration.
+func (cfg *JaegerReceiverCfg) InitFromViper(v *viper.Viper) (*JaegerReceiverCfg, error) {
+	return cfg, initFromViper(cfg, v, receiversRoot, jaegerEntry)
+}
+
+// OpenCensusReceiverCfg holds configuration for OpenCensus receiver.
+type OpenCensusReceiverCfg struct {
+	// Port is the port that the receiver will use
+	Port int `mapstructure:"port"`
+}
+
+// OpenCensusReceiverEnabled checks if the OpenCensus receiver is enabled, via a command-line flag, environment
+// variable, or configuration file.
+func OpenCensusReceiverEnabled(v *viper.Viper, cmdFlag string) bool {
+	return featureEnabled(v, cmdFlag, receiversRoot, opencensusEntry)
+}
+
+// NewOpenCensusReceiverCfg returns an instance of OpenCensusReceiverCfg with default values
+func NewOpenCensusReceiverCfg() *OpenCensusReceiverCfg {
+	opts := &OpenCensusReceiverCfg{
+		Port: 55678,
+	}
+	return opts
+}
+
+// InitFromViper returns a OpenCensusReceiverCfg according to the configuration.
+func (cfg *OpenCensusReceiverCfg) InitFromViper(v *viper.Viper) (*OpenCensusReceiverCfg, error) {
+	return cfg, initFromViper(cfg, v, receiversRoot, opencensusEntry)
+}
+
+// ZipkinReceiverCfg holds configuration for Zipkin receiver.
+type ZipkinReceiverCfg struct {
+	// Port is the port that the receiver will use
+	Port int `mapstructure:"port"`
+}
+
+// ZipkinReceiverEnabled checks if the Zipkin receiver is enabled, via a command-line flag, environment
+// variable, or configuration file.
+func ZipkinReceiverEnabled(v *viper.Viper, cmdFlag string) bool {
+	return featureEnabled(v, cmdFlag, receiversRoot, zipkinEntry)
+}
+
+// NewZipkinReceiverCfg returns an instance of ZipkinReceiverCfg with default values
+func NewZipkinReceiverCfg() *ZipkinReceiverCfg {
```

Comment:
Without `Default` in this name, the function surprisingly returns a default when one expects to do a full initialization.

Reply:
Good suggestion.